### PR TITLE
feat(frontend): Implement  aws-js-sdk crendentials to support IRSA for s3

### PR DESCRIPTION
**Pull request by ryansteakley**
**Description of your changes:**

#8502 details most of the changes and reasoning in a design document.

Removing most of the aws-helper code that pertains to grabbing the credentials from the ec2 instance metadata as this code is already handled as part of the `credentialProviderChain` that is being imported here.

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by ryansteakley** [comment]
@james-jwu @chensun @zijianjoy

**Comment by ryansteakley** [comment]
Looks like the prow job is running with node 12 installed, when it should move to using node 14? https://github.com/GoogleCloudPlatform/oss-test-infra/blob/b3128a944261abbbc3e23b7db974f95ea6c93192/prow/prowjobs/kubeflow/pipelines/kubeflow-pipelines-presubmits.yaml#L9

**Review by mvoitko — APPROVED** [review]

**Comment by ryansteakley** [comment]
Let me fix this unit-test looks like this is expected now since the prow has no credentials and it has exhausted all options. Missed this locally since my local env must have had some lingering creds somewhere.

File: frontend/server/minio-helper.ts [hunk 1]
```diff
@@ -39,22 +39,21 @@ export interface MinioClientOptionsWithOptionalSecrets extends Partial<MinioClie
 export async function createMinioClient(config: MinioClientOptionsWithOptionalSecrets) {
   if (!config.accessKey || !config.secretKey) {
     try {
-      if (await awsInstanceProfileCredentials.ok()) {
-        const credentials = await awsInstanceProfileCredentials.getCredentials();
```
Comment (zijianjoy) [inline frontend/server/minio-helper.ts:43]: In the design doc, you mention that we will fail back to existing solution if the IRSA doesn't work. It doesn't look like the case in this block. Can we retain the existing solution?

File: frontend/server/minio-helper.ts [hunk 2]
```diff
@@ -39,22 +39,21 @@ export interface MinioClientOptionsWithOptionalSecrets extends Partial<MinioClie
...
-        if (credentials) {
-          const {
-            AccessKeyId: accessKey,
-            SecretAccessKey: secretKey,
-            Token: sessionToken,
-          } = credentials;
-          return new MinioClient({ ...config, accessKey, secretKey, sessionToken });
-        }
-        console.error('unable to get credentials from AWS metadata store.');
+      const credentials = fromNodeProviderChain();
```
Comment (zijianjoy) [inline frontend/server/minio-helper.ts:42]: Is it possible to detect the source of object storage is AWS first, and perform IRSA only when it is AWS S3? (Pattern like "arn:aws:s3:::$mybucket/*") If it is generic S3 API, then we shouldn't use IRSA solution.

Comment (ryansteakley) [reply frontend/server/minio-helper.ts:42]: @zijianjoy  The code that used to be in this block was checking for AWS metadata store credentials. The fromNodeProviderChain() will try to access IRSA credentials then fallback and one of those would be AWS metadata store credentials which retains the exsiting functionality. 

If the option is preferred, then what I can do is call the existing https://github.com/kubeflow/pipelines/blob/834d966033ebb8284981e86b831c2bb086ef30ba/frontend/server/aws-helper.ts#L54 and have this IRSA code live as an if statement above the rest of the code in this method. i.e (if isS3Endpoint) ->my new code

Comment (zijianjoy) [reply frontend/server/minio-helper.ts:42]: Hello Ryan, thank you for replying to my comment, Sounds good on using `isS3Endpoint` , I would also suggest adding a comment to explain that which logic area is specific to AWS, and which logic is generic S3.

File: frontend/server/minio-helper.ts [hunk 3]
```diff
@@ -37,6 +38,25 @@ export interface MinioClientOptionsWithOptionalSecrets extends Partial<MinioClie
  * @param config minio client options where `accessKey` and `secretKey` are optional.
  */
 export async function createMinioClient(config: MinioClientOptionsWithOptionalSecrets) {
+  // This logic is AWS S3 specific
+  if (isS3Endpoint(config.endPoint)) {
```
Comment (surajkota) [inline frontend/server/minio-helper.ts:42]: is endpoint always set when running on an AWS by default (I dont remember us setting this as of now)?
or this becomes a required field now (this would be a breaking change)?

File: frontend/server/aws-helper.test.ts [hunk 4]
```diff
@@ -118,6 +118,10 @@ describe('isS3Endpoint', () => {
     expect(isS3Endpoint('s3.cn-north-1.amazonaws.com.cn')).toBe(true);
   });
 
+  it('checks a valid s3 fips GovCloud endpoint', () => {
```
Comment (surajkota) [inline frontend/server/aws-helper.test.ts:121]: +1

Can you add one for private link endpoint or does it have a similar format?

File: frontend/server/minio-helper.test.ts [hunk 5]
```diff
@@ -19,7 +19,6 @@ import { createMinioClient, isTarball, maybeTarball, getObjectStream } from './m
 import { V1beta1JobTemplateSpec } from '@kubernetes/client-node';
 
 jest.mock('minio');
-jest.mock('./aws-helper');
```
Comment (surajkota) [inline frontend/server/minio-helper.test.ts:22]: why is this removed? (I have no context on these tests, just curious)

File: frontend/server/minio-helper.ts (hunk 2 above)
Comment (surajkota) [reply frontend/server/minio-helper.ts]: @zijianjoy What is an example of generic S3 which has `s3://` in the source/URL?
https://github.com/kubeflow/pipelines/blob/master/frontend/server/handlers/artifacts.ts#L33

File: frontend/server/minio-helper.test.ts (hunk 5 above)
Comment (ryansteakley) [reply frontend/server/minio-helper.test.ts]: Was removed in the initial implementation since I had removed all the previous mocks that used aws-helper. Now that am reverting those changes back this will be necessary once again good catch.

File: frontend/server/aws-helper.test.ts (hunk 4 above)
Comment (ryansteakley) [reply frontend/server/aws-helper.test.ts:121]: Will add one for private-link endpoint, all amazon s3 endpoints follow the same format of having s3 prefix in the string and then amazonaws.com afterwards.

File: frontend/server/minio-helper.ts (hunk 3 above)
Comment (ryansteakley) [reply frontend/server/minio-helper.ts:42]: When `createMinioClient` is called in the `s3` case it passes in the aws `artifactsConfig` which contains endpoint, that resolves to the default value of `s3.amazonaws.com` or the env variable of `AWS_S3_ENDPOINT.` I'm not sure what you mean by breaking change?

Comment (surajkota) [reply frontend/server/minio-helper.ts:42]: https://github.com/kubeflow/pipelines/blob/88a9a31d86d34e61ac0cce9ed22f3455d91c5427/frontend/server/configs.ts#L128

Got it, thanks

File: frontend/server/minio-helper.ts [hunk 6]
```diff
@@ -37,6 +38,25 @@ export interface MinioClientOptionsWithOptionalSecrets extends Partial<MinioClie
...
+      }
+    } catch (err) {
+      console.error('Unable to get credentials from AWS credential provider chain: ', err);
+    }
+  }
+
+  // This logic is S3 generic
   if (!config.accessKey || !config.secretKey) {
     try {
       if (await awsInstanceProfileCredentials.ok()) {
```
Comment (surajkota) [inline frontend/server/minio-helper.ts:62]: This code is outdated and should be removed since there is an updated implementation for AWS S3. New implementation supports multiple ways to provide credentials including better security practices.

Better to discuss it in a separate PR

**Review by surajkota — APPROVED** [review]
Good work

**Comment by ryansteakley** [comment]
@zijianjoy Can you take another look?

**Comment by surajkota** [comment]
friendly reminder @zijianjoy since next release might be planned soon. We want to test this as part of distribution testing. Thanks

**Comment by jlyaoyuli** [comment]
Hi @surajkota @ryansteakley ,
This is Joe from KFP frontend, I will review the PR as soon as possible. Thank you!

**Comment by jlyaoyuli** [comment]
/lgtm
/approve
Thank you @ryansteakley and @surajkota for this implementation.
